src/cocaine-app/sched/__init__.py

```diff
@@ -0,0 +1,417 @@
+
+                    consumption = 100 / min(self.res_limits[job_type][res_type], 1)
+                    res_demand[(res_type, addr)] = consumption
+                else:
+                    if addr[0] not in storage.hosts and errors:
+                        errors.append('Host with address {} is not found in storage'.format(addr[0]))
+
+                    # Assume that FS is always 100% utilized
+                    res_demand[(Job.RESOURCE_FS, addr[0], addr[1])] = 100
+
+        return res_demand
+
+    def add_to_resource_stat(self, res_demand, job_id):
+        """
+        Add to saved resource representation resource consumption for the specified job. We could skip this and call
+        update_resource_stat after each iteration but this is rather expensive. So while the representation won't be
+        actual, let's keep it
+        :param res_demand: a dict where keys are tuples (res_type, node_addr\group_id, [fs_id])
+                            and values are consumption level
+        :param job_id: resource owner
+        """
+        for r in res_demand:
+            self.res[r].append((res_demand[r], job_id))
+
+    def get_resource_crossing(self, res_demand):
+        """
+        Uncovers inter-crossing of resources between running jobs and specified resources
+        :param resources: See convert_resource_representation
+        :return: list of (res_level, job_id)
+        """
+        crossing_jobs = []
+
+        for r in res_demand:
+            crossing_jobs.extend(self.res[r])
+
+        return crossing_jobs
+
+    def cancel_crossing_jobs(self, job_type, sched_params, crossing_jobs, demand):
+        """
+        Try to cancel jobs that are concurrent with a job we are trying to create
+        :param job_type: the type of the job we are trying to create
+        :param sched_params: scheduling params of the job we are trying to create
+        :param crossing_jobs:  see return of get_resource_crossing
+        :param demand: required resources (see convert_resource_representation)
+        :return: True if cancellation was successful, False otherwise
+        """
+        force = sched_params.get('force', False)
+
+        # XXX: these jobs should not be hard-coded. This should be present in config
+        STOP_ALLOWED_TYPES = (jobs.JobTypes.TYPE_RECOVER_DC_JOB,
+                              jobs.JobTypes.TYPE_COUPLE_DEFRAG_JOB,
+                              jobs.JobTypes.TYPE_TTL_CLEANUP_JOB
+                              )
+
+        job_ids = [v[1] for v in crossing_jobs]
+
+        existing_jobs = self.job_processor.job_finder.jobs(ids=job_ids, sort=False)
+        cancellable_jobs = []
+        for job in existing_jobs:
+
+            if self.job_processor.JOB_PRIORITIES[job.type] >= self.job_processor.JOB_PRIORITIES[job_type] and not force:
+                logger.info('Cannot stop job {} type {} since it has >= priority and no force'.format(job.id, job.type))
+                continue
```

Comment:
Fixed in https://github.com/yandex/mastermind/pull/417/commits/69574909fafa454bb1f62443cd1b2b78de089e76